File: backend/rorapp/serializers/faction.py
from django.contrib.auth.models import User
from rest_framework import serializers

from rorapp.models import Faction


class PlayerSerializer(serializers.ModelSerializer):
    class Meta:
        model = User
        fields = ["id", "username"]


class FactionPublicSerializer(serializers.ModelSerializer):
    player = PlayerSerializer(read_only=True)

    class Meta:
        model = Faction
        fields = ["id", "game", "player", "position", "card_count", "status"]
        read_only_fields = ["card_count"]


class FactionPrivateSerializer(serializers.ModelSerializer):
    """Only used for WebSocket messages"""
    player = PlayerSerializer(read_only=True)

    class Meta:
        model = Faction
        fields = ["id", "game", "player", "position", "treasury", "cards", "card_count", "status"]

File: frontend/classes/Faction.ts
export interface FactionData {
  id: number
  player: {
    id: number
    username: string
  }
  position: number
  card_count: number
  status_items: string[]
  display_name: string
}

class Faction {
  id: number
  player: {
    id: number
    username: string
  }
  position: number
  cardCount: number
  statusItems: string[]
  displayName: string

  constructor(data: FactionData) {
    this.id = data.id
    this.player = data.player
    this.position = data.position
    this.cardCount = data.card_count
    this.statusItems = data.status_items
    this.displayName = data.display_name
  }
}

export default Faction
